docs/ocis/adr/0029-grpc-in-kubernetes.md

```diff
@@ -0,0 +1,102 @@
+---
+title: "29. gRPC in Kubernetes"
+date: 2024-06-27T14:05:00+01:00
+weight: 29
+geekdocRepo: https://github.com/owncloud/ocis
+geekdocEditPath: edit/master/docs/ocis/adr
+geekdocFilePath: 0029-grpc-in-kubernetes.md
+---
+
+* Status: draft
+* Deciders: [@butonic](https://github.com/butonic)
+* Date: 2024-06-27
+
+## Context and Problem Statement
+
+[Scaling oCIS in kubernetes causes requests to fail
+#8589](https://github.com/owncloud/ocis/issues/8589), sometimes until affected services have manually been restarted. This can be observed by two symptoms:
+1. when a new pod is added it does not receive traffic
+2. when an pod is shut down existing clients still try to send requests to it
+
+To leverage the kubernetes pod state, we first used the go micro kubernetes registry implementation. When a pod fails the health or readyness probes, kubernetes will no longer
+- send traffic to the pod via the kube-proxy, which handles the ClusterIP for a service, 
+- list the pod in DNS responses when the ClusterIP is disabled by setting it to `none`
+When using the ClusterIP HTTP/1.1 requests will be routed to a working pod. 
+
+This nice setup starts to fail with long lived connections. The kube-proxy is connection based, causing requests with Keep-Alive to stick to the same pod for more than one request. Worse, HTTP/2 and in turn gRPC are multiplexing the connection. They will not pick up any changes to pods, explaining the symptomps:
+1. new pods will not be used because clients will reuse the existing gRPC connection
+2. gRPC clients will still try to send traffic to killed pods because they have not picked up that the pod was killed. Or the pod was killed a millisecond after the lookup was made.
+
+An addition to this problem are the health and readyness implementations of oCIS services not always reflecting the correct state of the service. One example is the storage-users service that returns ready `true` while runing a migration on startup.
```

Comment:
From what I know all /healthz and /readyz endpoints are hardcoded to true. Which is funny because the debug server might be up before the actual service server.